### PR TITLE
Ayon: deadline implementation

### PR DESCRIPTION
**Pull request by kalisp**
## Changelog Description
Quick implementation of deadline in Ayon. New Ayon plugin added for Deadline repository

## Additional info
Both implementation of OP and Ayon need to live together for a while, updated `GlobalJobPreLoad.py` to contain both. Simplest way chosen for first iteration.

!!! Known issue: Ayon Anatomy template don't contain `render` template as OP does and they should, imho.
Workaround it in for now in `submit_publish_job`, but still an issue in publish job in `Integrate`

## Testing notes:
1. deploy `modules/deadline/repository/custom` to DL repo
2. make sure that you have existing template name used in `tools/publish/template_name_profiles`
AnyDeskMSI_wbCWA1kmc4
(I used `publish` as `publish_render` is not yet set in Anatomy, but will be in the future)
3. run ayon mode and try to publish via DL (I was running `only_AE_DL` bundle in staging on ayon.dev)
4. run OP mode and try to publish via DL

### CODE REVIEW

**Comment by kalisp** [comment]
Will be gone for couple of days, but wanted to push semi-working implementation for comments. 
The main issue is missing `render` template in Ayon, that needs to be sorted out.

**Comment by kalisp** [comment]
Reworked hardcoded `render` template name in `Submit publish job`, it is using regular resolving of profile for template name.

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py [hunk 1]
```diff
@@ -398,6 +405,153 @@ def inject_openpype_environment(deadlinePlugin):
...
+        if all(add_kwargs.values()):
+            for key, value in add_kwargs.items():
+                args.extend(["--{}".format(key), value])
+        else:
+            raise RuntimeError((
+                "Missing required env vars: AVALON_PROJECT, AVALON_ASSET,"
+                " AVALON_TASK, AVALON_APP_NAME"
+            ))
+
+        os.environ["AVALON_TIMEOUT"] = "5000"
```
Comment (antirotor) [inline openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py:486]: This isn't probably needed in AYON as this sets timeout for mongodb connection if I am not mistaken.

**Review by antirotor — COMMENTED** [review]
code looks good so far, I need to test it.

File: openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py (hunk 1 above)
Comment (kalisp) [reply openpype/modules/deadline/repository/custom/plugins/GlobalJobPreLoad.py]: Ok, removed.

**Comment by mkolar** [comment]
So I'm hitting this on on the rendering job in the globaljobPreLoad

```
2023-08-02 13:56:49:  0: Loading Job's Plugin timeout is Disabled
2023-08-02 13:56:49:  0: SandboxedPlugin: Render Job As User disabled, running as current user 'milan'
2023-08-02 13:56:50:  0: Executing plugin command of type 'Initialize Plugin'
2023-08-02 13:56:50:  0: INFO: Executing plugin script 'C:\ProgramData\Thinkbox\Deadline10\workers\ygor\plugins\64ca43fcfa40e200764f8777\Nuke.py'
2023-08-02 13:56:51:  0: INFO: Plugin execution sandbox using Python version 3
2023-08-02 13:56:51:  0: INFO: About: Nuke Plugin for Deadline
2023-08-02 13:56:51:  0: INFO: The job's environment will be merged with the current environment before rendering
2023-08-02 13:56:51:  0: Done executing plugin command of type 'Initialize Plugin'
2023-08-02 13:56:51:  0: Start Job timeout is disabled.
2023-08-02 13:56:51:  0: Task timeout is disabled.
2023-08-02 13:56:51:  0: Loaded job: tst_bob_modelling_v018.nk - renderModellingMain (64ca43fcfa40e200764f8777)
2023-08-02 13:56:51:  0: Executing plugin command of type 'Start Job'
2023-08-02 13:56:51:  0: DEBUG: S3BackedCache Client is not installed.
2023-08-02 13:56:51:  0: INFO: Executing global asset transfer preload script 'C:\ProgramData\Thinkbox\Deadline10\workers\ygor\plugins\64ca43fcfa40e200764f8777\GlobalAssetTransferPreLoad.py'
2023-08-02 13:56:51:  0: INFO: Looking for legacy (pre-10.0.26) AWS Portal File Transfer...
2023-08-02 13:56:51:  0: INFO: Looking for legacy (pre-10.0.26) File Transfer controller in C:/Program Files/Thinkbox/S3BackedCache/bin/task.py...
2023-08-02 13:56:51:  0: INFO: Could not find legacy (pre-10.0.26) AWS Portal File Transfer.
2023-08-02 13:56:51:  0: INFO: Legacy (pre-10.0.26) AWS Portal File Transfer is not installed on the system.
2023-08-02 13:56:51:  0: INFO: Executing global job preload script 'C:\ProgramData\Thinkbox\Deadline10\workers\ygor\plugins\64ca43fcfa40e200764f8777\GlobalJobPreLoad.py'
2023-08-02 13:56:51:  0: PYTHON: *** GlobalJobPreload start ...
2023-08-02 13:56:51:  0: PYTHON: >>> Getting job ...
2023-08-02 13:56:51:  0: PYTHON: >>> Injecting Ayon environments ...
2023-08-02 13:56:51:  0: PYTHON: --- Ayon executable: C:\Users\milan\AppData\Local\ynput\ayon\app\AYON 1.0.0-beta.1\ayon.exe
2023-08-02 13:56:51:  0: PYTHON: Traceback (most recent call last):
2023-08-02 13:56:51:  0: PYTHON:   File "C:\ProgramData\Thinkbox\Deadline10\workers\ygor\plugins\64ca43fcfa40e200764f8777\GlobalJobPreLoad.py", line 432, in inject_ayon_environment
2023-08-02 13:56:51:  0: PYTHON:     raise RuntimeError("Missing env var in job properties "
2023-08-02 13:56:51:  0: PYTHON: RuntimeError: Missing env var in job properties AYON_BUNDLE_NAME
2023-08-02 13:56:51:  0: PYTHON: !!! Injection failed.
2023-08-02 13:56:51:  0: 'C:\Users\milan\AppData\Local\Thinkbox\Deadline10\pythonAPIs\2023-04-17T214216.0000000Z' already exists. Skipping extraction of PythonSync.
2023-08-02 13:56:51:  0: Done executing plugin command of type 'Start Job'
```

File: openpype/modules/deadline/abstract_submit_deadline.py [hunk 2]
```diff
@@ -397,6 +397,15 @@ def update(self, data):
         for key, value in data.items():
             setattr(self, key, value)
 
+    def add_render_job_env_var(self):
+        """Check if in OP or AYON mode and use appropriate env var."""
+        if os.environ.get("USE_AYON_SERVER") == '1':
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/abstract_submit_deadline.py:402]:
```suggestion
        if AYON_SERVER_ENABLED:
```

Requires `from openpype import AYON_SERVER_ENABLED`

File: openpype/modules/deadline/plugins/publish/submit_maya_remote_publish_deadline.py [hunk 3]
```diff
@@ -114,11 +114,16 @@ def get_job_info(self):
         environment["AVALON_TASK"] = instance.context.data["task"]
         environment["AVALON_APP_NAME"] = os.environ.get("AVALON_APP_NAME")
         environment["OPENPYPE_LOG_NO_COLORS"] = "1"
-        environment["OPENPYPE_REMOTE_JOB"] = "1"
         environment["OPENPYPE_USERNAME"] = instance.context.data["user"]
         environment["OPENPYPE_PUBLISH_SUBSET"] = instance.data["subset"]
         environment["OPENPYPE_REMOTE_PUBLISH"] = "1"
 
+        if os.environ.get("USE_AYON_SERVER") == '1':
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/plugins/publish/submit_maya_remote_publish_deadline.py:121]:
```suggestion
        if AYON_SERVER_ENABLED:
```

File: openpype/modules/deadline/plugins/publish/submit_publish_job.py [hunk 4]
```diff
@@ -203,13 +198,25 @@ def _submit_deadline_post_job(self, instance, job, instances):
...
             "AVALON_TASK": instance.context.data["task"],
             "OPENPYPE_USERNAME": instance.context.data["user"],
-            "OPENPYPE_PUBLISH_JOB": "1",
-            "OPENPYPE_RENDER_JOB": "0",
-            "OPENPYPE_REMOTE_JOB": "0",
             "OPENPYPE_LOG_NO_COLORS": "1",
             "IS_TEST": str(int(is_in_tests()))
         }
 
+        if os.environ.get("USE_AYON_SERVER") == '1':
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/plugins/publish/submit_publish_job.py:205]:
```suggestion
        if AYON_SERVER_ENABLED:
```

File: openpype/modules/deadline/repository/custom/plugins/Ayon/Ayon.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+                if paddingSize > 0:
+                    padding = StringUtils.ToZeroPaddedString(frame,
+                                                             paddingSize,
+                                                             False)
+                else:
+                    padding = str(frame)
+                arguments = arguments.replace(frameMatch.Groups[0].Value,
+                                              padding)
+            else:
+                break
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/repository/custom/plugins/Ayon/Ayon.py:145]:
```suggestion
            if not frameMatch.Success:
                break
            paddingSize = int(frameMatch.Groups[1].Value)
            if paddingSize > 0:
                padding = StringUtils.ToZeroPaddedString(
                    frame, paddingSize, False)
            else:
                padding = str(frame)
            arguments = arguments.replace(
                frameMatch.Groups[0].Value, padding)
```

File: openpype/modules/deadline/plugins/publish/submit_maya_remote_publish_deadline.py [hunk 6]
```diff
@@ -114,11 +115,16 @@ def get_job_info(self):
...
         environment["OPENPYPE_USERNAME"] = instance.context.data["user"]
         environment["OPENPYPE_PUBLISH_SUBSET"] = instance.data["subset"]
         environment["OPENPYPE_REMOTE_PUBLISH"] = "1"
 
+        if AYON_SERVER_ENABLED:
+            environment["AYON_REMOTE_PUBLISH"] = "1"
+        else:
+            environment["OPENPYPE_REMOTE_PUBLISH"] = "1"
+
+
```
Comment (iLLiCiTiT) [inline openpype/modules/deadline/plugins/publish/submit_maya_remote_publish_deadline.py:127]:
```suggestion

```

**Comment by kalisp** [comment]
Fixed issue with missing `AYON_BUNDLE_NAME` in Nuke.

**Review by mkolar — APPROVED** [review]
Tested with Maya and Nuke. All good!

**Pull request merged**